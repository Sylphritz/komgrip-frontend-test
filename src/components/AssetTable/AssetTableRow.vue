<template>
  <tr>
    <td class="text-center" data-test="asset-table-row-rank">{{ asset.rank }}.</td>
    <td data-test="asset-table-row-name">{{ asset.name }}</td>
    <td class="text-center" data-test="asset-table-row-symbol">{{ asset.symbol }}</td>
    <td class="text-end" data-test="asset-table-row-supply">{{ supply }} / {{ maxSupply }}</td>
    <td class="text-end" data-test="asset-table-row-price">{{ price }}</td>
    <td
      class="text-center"
      :class="{ 'text-success': rawChange > 0, 'text-danger': rawChange < 0 }"
      data-test="asset-table-row-change"
    >
      {{ changePercentage }}%
    </td>
  </tr>
</template>

<script setup lang="ts">
import { computed, type PropType } from 'vue'
import type { Asset } from '@/api/coincap/types'

const props = defineProps({
  asset: {
    type: Object as PropType<Asset>,
    required: true
  }
})

const supply = computed(() => parseInt(props.asset.supply).toLocaleString())
const maxSupply = computed(() =>
  props.asset.maxSupply
    ? parseFloat(props.asset.maxSupply).toLocaleString('en-US', {
        useGrouping: true
      })
    : 'NO LIMIT'
)
const price = computed(() =>
  parseFloat(props.asset.priceUsd).toLocaleString('en-US', {
    useGrouping: true,
    style: 'currency',
    currency: 'USD'
  })
)
const rawChange = computed(() => parseFloat(props.asset.changePercent24Hr))
const changePercentage = computed(
  () => `${rawChange.value > 0 ? '+' : ''}${rawChange.value.toFixed(2)}`
)
</script>
