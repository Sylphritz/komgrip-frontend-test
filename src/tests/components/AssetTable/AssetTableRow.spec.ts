import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import AssetTableRow from '@/components/AssetTable/AssetTableRow.vue'
import { MOCK_ASSETS_RESPONSE } from '@/mocks/api/assetsMock'

describe('AssetTableRow', () => {
  it('renders successfully and matches the snapshot', () => {
    const wrapper = mount(AssetTableRow, {
      shallow: true,
      props: {
        asset: MOCK_ASSETS_RESPONSE.data[0]
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders the displayed values in the correct format', () => {
    const asset = MOCK_ASSETS_RESPONSE.data[0]
    const wrapper = mount(AssetTableRow, {
      shallow: true,
      props: {
        asset
      }
    })

    expect(wrapper.find('[data-test="asset-table-row-rank"]').element.textContent).toBe('1.')
    expect(wrapper.find('[data-test="asset-table-row-name"]').element.textContent).toBe('Bitcoin')
    expect(wrapper.find('[data-test="asset-table-row-symbol"]').element.textContent).toBe('BTC')
    expect(wrapper.find('[data-test="asset-table-row-supply"]').element.textContent).toBe(
      '17,193,925 / 21,000,000'
    )
    expect(wrapper.find('[data-test="asset-table-row-price"]').element.textContent).toBe(
      '$6,929.82'
    )
    expect(wrapper.find('[data-test="asset-table-row-change"]').element.textContent?.trim()).toBe(
      '-0.81%'
    )
  })

  it('renders the price change in green when the value is positive', () => {
    const asset = MOCK_ASSETS_RESPONSE.data[0]
    const wrapper = mount(AssetTableRow, {
      shallow: true,
      props: {
        asset: {
          ...asset,
          changePercent24Hr: '1.68'
        }
      }
    })

    expect(
      wrapper
        .find('[data-test="asset-table-row-change"]')
        .element.classList.contains('text-success')
    ).toBe(true)
  })

  it('renders the price change in red when the value is negative', () => {
    const asset = MOCK_ASSETS_RESPONSE.data[0]
    const wrapper = mount(AssetTableRow, {
      shallow: true,
      props: {
        asset
      }
    })

    expect(
      wrapper.find('[data-test="asset-table-row-change"]').element.classList.contains('text-danger')
    ).toBe(true)
  })

  it('renders the price change in regular text color when the value is 0', () => {
    const asset = MOCK_ASSETS_RESPONSE.data[0]
    const wrapper = mount(AssetTableRow, {
      shallow: true,
      props: {
        asset: {
          ...asset,
          changePercent24Hr: '0'
        }
      }
    })

    expect(
      wrapper.find('[data-test="asset-table-row-change"]').element.classList.contains('text-danger')
    ).not.toBe(true)
    expect(
      wrapper
        .find('[data-test="asset-table-row-change"]')
        .element.classList.contains('text-success')
    ).not.toBe(true)
  })
})
