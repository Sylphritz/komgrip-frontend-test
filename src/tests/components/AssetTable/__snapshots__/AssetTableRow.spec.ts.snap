// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AssetTableRow > renders successfully and matches the snapshot 1`] = `
<tr>
  <td
    class="text-center"
    data-test="asset-table-row-rank"
  >
    1.
  </td>
  <td
    data-test="asset-table-row-name"
  >
    Bitcoin
  </td>
  <td
    class="text-center"
    data-test="asset-table-row-symbol"
  >
    BTC
  </td>
  <td
    class="text-end"
    data-test="asset-table-row-supply"
  >
    17,193,925 / 21,000,000
  </td>
  <td
    class="text-end"
    data-test="asset-table-row-price"
  >
    $6,929.82
  </td>
  <td
    class="text-center text-danger"
    data-test="asset-table-row-change"
  >
    -0.81% 
  </td>
</tr>
`;
